### PR TITLE
Bitonic sort on GPU (OpenCL)

### PR DESCRIPTION
**Pull request by margaritiko**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Added OpenCl support and bitonic sort algorithm, which can be used for sorting integer types of data in single column. Needs to be build with flag `-DENABLE_OPENCL=1`. For using bitonic sort algorithm instead of others you need to set `bitonic_sort` for Setting's option `special_sort` and make sure that OpenCL is available.

### CODE REVIEW

File: src/Common/BitonicSort.cl [hunk 1]
```diff
@@ -0,0 +1,6 @@
+__kernel void bitonic_sort(__global float *A, __global float *B, __global float *C, int N) {
+  int i = get_global_id(0);
+  if (i < N) {
+    C[i] = A[i] + B[i];
+  }
+}
```
Comment (4ertus2) [inline src/Common/BitonicSort.cl:6]: no empty line

File: src/Common/BitonicSort.cl [hunk 2]
```diff
@@ -0,0 +1,6 @@
+__kernel void bitonic_sort(__global float *A, __global float *B, __global float *C, int N) {
```
Comment (4ertus2) [inline src/Common/BitonicSort.cl:1]: - brace on it's own line
- spaces after *

File: src/Common/BitonicSort.cl [hunk 3]
```diff
@@ -0,0 +1,6 @@
...
+  int i = get_global_id(0);
```
Comment (4ertus2) [inline src/Common/BitonicSort.cl:2]: tab = 4 spaces

File: src/Common/BitonicSort.cl [hunk 4]
```diff
@@ -0,0 +1,6 @@
...
+  if (i < N) {
```
Comment (4ertus2) [inline src/Common/BitonicSort.cl:3]: no braces for simple one-line statements

File: src/Common/OpenCLBasics.cpp [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+
+static void checkError(cl_int error)
+{
+    if (error != CL_SUCCESS)
+        throw std::runtime_error("OpenCL error " + opencl_error_to_str(error) + ".");
+}
+
+
+static void* throwExceptionWithMessage(String message)
```
Comment (4ertus2) [inline src/Common/OpenCLBasics.cpp:98]: space before *

File: src/Common/BitonicSort.h [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+#include <map>
+#include <type_traits>
+
+#include <ext/bit_cast.h>
+#include <Core/Types.h>
+#include <Core/Defines.h>
+
+#include "OpenCLBasics.cpp"
+
+#define CL_USE_DEPRECATED_OPENCL_1_2_APIS
```
Comment (4ertus2) [inline src/Common/BitonicSort.h:27]: It does not make sence if definaed after `#include <CL/cl.h>`. We do not need it by default and should use OpenCL version from cmake to detect if it's old.

File: src/Columns/ColumnVector.cpp [hunk 7]
```diff
@@ -135,6 +144,32 @@ void ColumnVector<T>::getPermutation(bool reverse, size_t limit, int nan_directi
     }
     else
     {
+
+        /// If bitonic sort if specified as preferred
+#if defined (BITONIC_SORT_PREFERRED) && defined (BITONIC_SORT_IMPL)
+        T* pairsData = new T[s];
```
Comment (4ertus2) [inline src/Columns/ColumnVector.cpp:150]: It's not OK to make new/delete ourself. If there would and exception inside we'll get memory leak here. Make an std::vector<T>, resize it by s, and use `data()` for row array.

File: src/Columns/ColumnVector.cpp [hunk 8]
```diff
@@ -135,6 +144,32 @@ void ColumnVector<T>::getPermutation(bool reverse, size_t limit, int nan_directi
     }
     else
     {
+
+        /// If bitonic sort if specified as preferred
+#if defined (BITONIC_SORT_PREFERRED) && defined (BITONIC_SORT_IMPL)
```
Comment (4ertus2) [inline src/Columns/ColumnVector.cpp:149]: move this block in own function

File: src/Common/BitonicSort.h (hunk 6 above)
Comment (margaritiko) [reply src/Common/BitonicSort.h]: Fixed with checking `CL_VERSION_2_0`

File: src/Columns/ColumnVector.cpp (hunk 7 above)
Comment (margaritiko) [reply src/Columns/ColumnVector.cpp]: Resolved by using vectors and moving all this data manipulation to the method in BitonicSort class

File: src/Common/BitonicSort.h [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+#include <Core/Defines.h>
+#include <Common/PODArray.h>
+#include <Columns/ColumnsCommon.h>
+
+#include "oclBasics.cpp"
+
+class BitonicSort
+{
+public:
+    static BitonicSort& getInstance()
```
Comment (4ertus2) [inline src/Common/BitonicSort.h:36]: space before &

File: src/Common/BitonicSort.h [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+
+        for (UInt32 i = 0; i < s; ++i)
+        {
+            pairsContent[i] = data[i];
+            pairsIndices[i] = i;
+        }
+
+        bool result = sort(pairsContent.data(), pairsIndices.data(), s, sortAscending);
+
+        if (!result) return false;
```
Comment (4ertus2) [inline src/Common/BitonicSort.h:101]: return fale at own line

File: src/Common/BitonicSort.h [hunk 11]
```diff
@@ -0,0 +1,349 @@
...
+
+            OCL::checkError(error);
+
+            std::vector<char> log(logSize);
+            clGetProgramBuildInfo(program, configuration.device, CL_PROGRAM_BUILD_LOG, logSize, log.data(), nullptr);
+
+            OCL::checkError(error);
+            OCL::throwExceptionWithMessage(log.data());
+
+            exit(EXIT_FAILURE);
```
Comment (4ertus2) [inline src/Common/BitonicSort.h:316]: It should be an exception here.

File: src/Common/oclBasics.cpp [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+#undef CASE_CL_CONSTANT
+}
+
+
+static void checkError(cl_int error)
+{
+    if (error != CL_SUCCESS)
+        throw std::runtime_error("OpenCL error " + opencl_error_to_str(error) + ".");
```
Comment (4ertus2) [inline src/Common/oclBasics.cpp:97]: We should throw our Exception instead. Add new code entry to Common/ErrorCodes.cpp for it. I.e. `OPENCL_ERROR`

File: src/Columns/ColumnVector.cpp [hunk 13]
```diff
@@ -135,6 +139,14 @@ void ColumnVector<T>::getPermutation(bool reverse, size_t limit, int nan_directi
     }
     else
     {
+        /// If bitonic sort if specified as preferred
+#if defined (BITONIC_SORT_PREFERRED)
+        if (BitonicSort::getInstance().sort(data, res, true))
+        {
```
Comment (4ertus2) [inline src/Columns/ColumnVector.cpp:145]: Do not need these brases for one line `return`

**Comment by qoega** [comment]
We need tests for this feature

File: src/Common/oclBasics.cpp [hunk 14]
```diff
@@ -0,0 +1,378 @@
...
+        OCL::checkError(error);
+
+        return commandQueue;
+    }
+
+
+    static cl_program makeProgram(const char * file_name, cl_context context,
+                                  cl_device_id device_id, const Settings & settings)
+    {
+        FILE * file = fopen(file_name, "rb");
```
Comment (4ertus2) [inline src/Common/oclBasics.cpp:269]: This logic does not work. We should not use resources. Define kernels as constant strings.

**Review by 4ertus2 — CHANGES_REQUESTED** [review]

File: src/Common/oclBasics.cpp [hunk 15]
```diff
@@ -0,0 +1,378 @@
...
+struct OCL
+{
+
+    /**
+     * Structure which represents the most essential settings of common OpenCl entities.
+     */
+    struct Settings
+    {
+        // Platform info
+        cl_uint numberOfPlatformEntries;
```
Comment (4ertus2) [inline src/Common/oclBasics.cpp:43]: Here and further. Coding style for variables: `numberOfPlatformEntries` -> `number_of_platform_entries`.

File: src/Core/SettingsCollection.cpp [hunk 16]
```diff
@@ -485,6 +485,12 @@ void SettingURI::deserialize(ReadBuffer & buf, SettingsBinaryFormat)
 IMPLEMENT_SETTING_ENUM(LoadBalancing, LOAD_BALANCING_LIST_OF_NAMES, ErrorCodes::UNKNOWN_LOAD_BALANCING)
 
 
+#define SPECIAL_SORT_ALGORITHM_NAMES(M) \
+    M(NOT_SPECIFIED, "not_specified") \
+    M(BITONIC_SORT, "bitonic_sort")
```
Comment (4ertus2) [inline src/Core/SettingsCollection.cpp:490]: Rename to `opencl_bitonic`. We could have cuda variant in future. And word `sort` is not needed cause it apears in setting name.

**Comment by 4ertus2** [comment]
Related #10898
I've tested current version with minimal fixes. It works.

File: src/Common/oclBasics.cpp (hunk 15 above)
Comment (margaritiko) [reply src/Common/oclBasics.cpp]: Renamed all variables ✔️

File: src/Columns/ColumnVector.cpp [hunk 17]
```diff
@@ -135,6 +137,10 @@ void ColumnVector<T>::getPermutation(bool reverse, size_t limit, int nan_directi
     }
     else
     {
+        /// If bitonic sort if specified as preferred than `nan_direction_hint` equals specific value 42.
+        if (nan_direction_hint == 42 && BitonicSort::getInstance().sort(data, res, !reverse))
```
Comment (4ertus2) [inline src/Columns/ColumnVector.cpp:141]: It's OK for now, but it's better to make enum and describe it in nan_direction_hint comment in IColumn.h. I'll make it later.

**Review by 4ertus2 — APPROVED** [review]

**Comment by paolovolterra** [comment]
Great
Have I to modify config.xml  to use it?

**Comment by alexey-milovidov** [comment]
@paolovolterra The implementation was removed, because no one has demonstrated the advantages.

**Pull request merged**